MetaMask/metamask-extension
# chore: update `@trezor/connect-web` to `v9.2.2`

## **Description**

The user experience for trezor connection on mv2 is this:
1. Click through the hardware wallet connect screen for trezor
2. Trezor popup tab opens
3. User can connect the trezor device without having to install other software

Before this PR, on mv3, the flow was:
1. Click through the hardware wallet connect screen for trezor
2. Trezor popup tab opens
3. User has to install the trezor bridge
4. Once installed, the user can connect their trezor device to MetaMask

The solution to the problem required two parts.

First, we needed the `connect-web` module (called `TrezorConnectSDK` in the offscreen trezor file) to specify the `env` variable communicated to the trezor connect iframe and popup to be "webextension" and not just "web". This is because the iframe/popup code assumes that if the iframe origin does not match the origin of popup (e.g. metamask vs connect.trezor.io), and if environment is not "webextension", then webusb cannot be available: https://github.com/trezor/trezor-suite/blob/bb2e075024c8d8316fa016b7b20a0421b1a1f7d0/packages/connect-iframe/src/connectSettings.ts#L74-L84.

In theory, this could be set when the trezor module is initialized, but if the `env` is set to "webextension" in the `connect-web` code, then it attempts to open the popup via chrome apis that are not available in the offscreen document. So the solution is a patch which just tells the iframe and the popup that the environment is "webextension", while leave the environment setting within the trezor code that runs in the offscreen context as "web".

With that fix in place, there was another problem of metamask being able to connect to the webusb device and the popup being able to read that connection (or vice-versa). If the `requestDevice` call happens from the popup, then the webusb permission will be granted to the trezor.io origin, and metamask is unable to communicate with those devices. However, if the `requestDevice` call happens from metamask, then the devices can be communicated from metamask to the popup, and metamask (in particular the offscreen document) can communicate directly with the devices. Specifically, this code https://github.com/trezor/trezor-suite/blob/bb2e075024c8d8316fa016b7b20a0421b1a1f7d0/packages/transport/src/api/usb.ts#L98, when run within the offscreen document, returns null when `requestDevice` was called from the trezor popup, but returns a correctly populated array if `requestDevice` was called from MetaMask. To fix this, we can call `requestDevice` explicitly within the MetaMask ui flow, instead of having the trezor popup call it for us

Finally, and independent of the above two fixes, this PR updates trezor connect packages, which was necessary to get trezor working properly with Snow.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23763?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Build this branch with `yarn start:mv3`
2. Go through the regular trezor connect flow
3. Trezor should connect and sign as normal

## **Screenshots/Recordings**


https://github.com/MetaMask/metamask-extension/assets/7499938/29533595-f16f-4492-be21-f9bc3698fe70



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
